azure-go-sdk-conceptual/azure-sdk-go-qs-vm.md

```diff
@@ -0,0 +1,307 @@
+
+        template, err := readJSON(templateFile)
+        if err != nil {
+                return resources.DeploymentExtended{}, err
+        }
+        params, err := readJSON(parametersFile)
+        if err != nil {
+                return resources.DeploymentExtended{}, err
+        }
+
+        deploymentFuture, err := deploymentsClient.CreateOrUpdate(
+                ctx,
+                resourceGroupName,
+                deploymentName,
+                resources.Deployment{
+                        Properties: &resources.DeploymentProperties{
+                                Template:   template,
+                                Parameters: params,
+                                Mode:       resources.Incremental,
+                        },
+                },
+        )
+        if err != nil {
+                log.Fatalf("Failed to create deployment: %v", err)
+        }
+        //...
+```
+
+We'll start with the `readJSON` function. It just loads up the JSON contained within the specified file, so the details are skipped here. The important element is that it returns a `*map[string]interface{}`, the value taken by both the `Template` and `Parameters` fields of the `DeploymentProperties` struct. Again, structs representing properties of Azure services almost exclusively take pointers.
+
+Although this looks more complicated than creating the resource group, it follows the same pattern: A new client is created, it is given the ability to authenticate with Azure, and then a method is called. The method even has the same name (`CreateOrUpdate`) as the corresponding method for resource groups. This is a pattern that you will see again and again in the SDK. Methods which perform similar work tend to have the same name.
+
+The biggest difference comes in the return value of the `CreateOrUpdate()` method. This is a `Future` object, which follows the [future design pattern](https://en.wikipedia.org/wiki/Futures_and_promises) and represents a long-running operation in Azure that you may want to occasionally poll while performing other work.
+
+```golang
+        //...
+        err = deploymentFuture.Future.WaitForCompletion(ctx, deploymentsClient.BaseClient.Client)
+        if err != nil {
+                log.Fatalf("Error while waiting for deployment creation: %v", err)
+        }
+        return deploymentFuture.Result(deploymentsClient)
+}
+```
+
+For our purposes, the best thing to do is to wait for the operation to complete since there is no other work to perform. Waiting on a future requires both a [context object](https://blog.golang.org/context) and the client which is responsible for managing the operation represented by the Future itself. Note that even with an error possibly occuring while waiting for completion, this is _distinct_ from the error which could be a natural part of resolving the operation on the server side, which is returned as part of the `Result()` call.
+
+### Obtaining the assigned IP address

```

Comment:
This section needs to be rewritten. There is actually a `services` module for public IP, but it is NOT picked up by godoc, and so I didn't know it existed. Seems bad.